Improvements

shinenelson opened the pull request:
This is almost like a whole re-factor of the bot.
The commit messages should be self-explanatory.

* Change birthdays database to YAML / JSON
The YAML module can load and parse JSON files, so that's an added benefit. I initially thought of using only JSON since the configuration file was JSON too. But since YAML is a more easier-to-understand format for everyone, so I thought I'd add support for that as well.

* Configurize `db_path` and `mention`
   * The `birthdays.txt` was being hard-coded and loaded from the code. This makes the bot less-flexible in terms of the database filename. Since I was re-factoring the code, I thought why not. Configurizing the database filename would make it more flexible. But the database file should still be parse-able by the YAML module, so that's there.
   * Some people prefer wishing their team members with real names while others like mentioning them and spam them with mentions all day. Mentioning is configurable too. This too depends on what values are loaded from the database. If the database has real names and the `mention` flag is `true`, the purpose is defeated.

* Renaming `configurations.json` to `configurations.json.example` - because that's what it is, only an example placeholder file, not an actual configuration file.

* The bot has been optimized to work with the above changes and the README file has been updated with more information with regard on how to configure the bot and use it.

shinenelson committed:
Change birthdays database to YAML / JSON

Parsing plain text files requires more processing than
structured file types like YAML / JSON.
YAML support is built-into ruby > 1.9.3.
The YAML module can easily load a JSON file too.

Why was the year field removed?
Birthdays are recurring. You need to wish people every year.
Why would you want to torment people with the notion of them getting old? :stuck_out_tongue_winking_eye:

shinenelson committed:
Configurize db_path for the birthdays database

shinenelson committed:
Rename configurations.json to configuration.json.example

because that is what it is - an example / placeholder configuration file

shinenelson committed:
Optimize bot to work with {hash}ed birthdays

Supports multiple birthdays from arrays.
Earlier version sent multiple wishes for multiple birthdays,
this version appends the users to a single array
and sends out a single birthday wish message.

shinenelson committed:
Configurize mention'ing team members

You can now configure the birthday database
with your team memebers' usernames or real names.
The bot will mention the users based on this parameter.

shinenelson committed:
Update README.md with more detailed instructions

mmachado95 commented:
@teresalazar13

jbernardo95 commented on lib/birthday_reader.rb:
```diff
@@ -1,12 +1,11 @@
-        birthdays << line.chomp.split
-      end
-      file.close
+	bday_list = YAML.load_file(file_path)
```
The purpose of this class is to read birthdays only, I don't think we should return only the birthdays according to `Time.now`, instead we should return all of them.

shinenelson replied:
The primary reason I switched to a structured format for the database was to drill down to the exact date to fetch all the birthdays on the day. What would be the use case to have return all the birthdays together? It would just use more memory (for no particular gain, whatsoever) IMHO.

jbernardo95 replied:
I am not against the structuring.

What I am saying is that the purpose of the function is to read birthdays. It is not to read and filter. That is why I think we should not filter the birthdays here.

And because we are just managing strings of birthdays I guess memory is not a concern.

jbernardo95 commented on configurations.json:
```diff
@@ -1,4 +1,5 @@
 {
+  "db_path":	      "birthdays.yaml",
```
I think it makes more sense to change `db_path` to `birthdays_path`

jbernardo95 commented on lib/birthday_bot.rb:
```diff
@@ -11,17 +11,28 @@ def initialize()
-                                                 username: @config.bot_name,
-                                                 text: message,
-                                                 icon_emoji: @config.bot_emoji }.to_json)
+    puts "Checking who was born today"
```
I think we should keep `Time.now` in the log

shinenelson replied:
Ah, you wanted it in the log. The way I handle my `cron`s are, I prepend `date` before my executing command. But I don't know if this works with Heroku's scheduler (does it?). Else I'll just put the date back in the log.

jbernardo95 commented on lib/birthday_bot.rb:
```diff
@@ -11,17 +11,28 @@ def initialize()
+           end
+           users += " and <@#{ birthdays[i+1] }> "
+        end
       end
```
I think we can simplify this:

```
puts "#{birthdays.count} were born today"

today = Time.now
birthdays_today = birthdays[today.month][today.day]

unless birthdays_today.empty?
  users = build_user_list(birthdays_today)
  message = "#{users} #{@config.greeting_message}"
  HTTParty.post(@config.slack_url, body: { channel: @config.channel_name,
    username: @config.bot_name,
    text: message,
    icon_emoji: @config.bot_emoji }.to_json)
end
 
# ...

def build_user_list(birthdays)
  if birthdays.size == 1
    birthdays.first
  else
    users = birthdays.take(birthdays.count - 1).join(", ")
    users += " and #{birthdays[birthdays.count - 1]}" if birthdays.count > 1
  end
end
```

What do you think ?

jbernardo95 replied:
@shinenelson Did you take a look at this ?

(mentions are missing here)

jbernardo95 commented on lib/birthday_bot.rb:
```diff
@@ -35,4 +35,13 @@ def start!
+    else
+      name
+    end
+  end
```
Can you give an example where disabling mentions is useful ?

shinenelson replied:
The earlier version of the database stored the real names of users. The bot wished users with their real names. So, I thought that could be a valid use-case

But like my commit message said, some people might prefer wishing their team-mates by their real names (probably for more family-togetherness) and/or not choosing to spam the users with mentions on the wishes (but that wil probably be taken care of by other team-mates, anyway).

jbernardo95 replied:
Makes sense, thanks !

jbernardo95 reviewed (changes requested):
@shinenelson Awesome PR ! :) The birthday settings with json/yaml is a really nice addition.

Left a couple of comments for you to take a loo.

shinenelson committed:
Rename db_path to birthdays_path (Resolves https://github.com/jeKnowledge/slack-birthday-bot/pull/9#discussion_r120021476)

shinenelson committed:
Log Time.now

shinenelson committed:
change date keys to string

The YAML module parses keys with a leading 0 as octcal digits,
invalidating 08 and 09. Parsing them as strings would be safer

shinenelson committed:
handle empty / nil YAML.load_file

shinenelson commented:
I'll leave a friendly reminder here that this pull request is still un-merged.

Also, I've handled 2 new issues that I came across.
1. `YAML` parses values with leading 0 as octal digits, it would be safer if we parsed the dates as strings than just integers.
2. Handled an exception with empty array after `YAML.load`.

mmachado95 commented:
@jbernardo95

jbernardo95 commented on lib/birthday_bot.rb:
```diff
@@ -3,26 +3,45 @@
-                                                 username: @config.bot_name,
-                                                 text: message,
-                                                 icon_emoji: @config.bot_emoji }.to_json)
+    puts "Checking who was born today ( #{ Time.now.to_s } )"
```
`puts "Checking who was born today (#{ Time.now.to_s })"` remove the spaces.

shinenelson replied:
I felt it would be more legible with the spaces.

jbernardo95 commented on lib/birthday_bot.rb:
```diff
@@ -3,26 +3,45 @@
+                                               icon_emoji: @config.bot_emoji }.to_json)
+    else
+      puts "Today is a day that no one was born"
     end
```
Have you looked at my previous comment with a suggestion to refactor this function ?

```
puts "#{birthdays.count} were born today"

today = Time.now
birthdays_today = birthdays[today.month][today.day]

unless birthdays_today.empty?
  users = build_user_list(birthdays_today)
  message = "#{users} #{@config.greeting_message}"
  HTTParty.post(@config.slack_url, body: { channel: @config.channel_name,
    username: @config.bot_name,
    text: message,
    icon_emoji: @config.bot_emoji }.to_json)
end
 
# ...

def build_user_list(birthdays)
  if birthdays.size == 1
    birthdays.first
  else
    users = birthdays.take(birthdays.count - 1).join(", ")
    users += " and #{birthdays[birthdays.count - 1]}" if birthdays.count > 1
  end
end
```

jbernardo95 reviewed (changes requested):
@shinenelson Sorry for the late reply I didn't get any notifications with updates on this.

I left a few extra comments for you to take a look.

Next time tag me using `@jbernardo95` to make sure I see the notifications.

shinenelson committed:
remove spaces (Resolves https://github.com/jeKnowledge/slack-birthday-bot/pull/9#discussion_r130992162)

shinenelson committed:
move birthday list filter to birthday_bot from birthday_reader

shinenelson committed:
add today = Time.now (Resolves https://github.com/jeKnowledge/slack-birthday-bot/pull/9#discussion_r120022086)

shinenelson committed:
def build_user_list

shinenelson committed:
[docs] replace db_path with birthdays_path

shinenelson committed:
[readme] fix URL in contributors' list

shinenelson committed:
[refactor] remove httparty dependency

HTTParty is the only gem dependency of the bot.
I thought it was redundant to maintain a single
dependency especially when it was achievable with
the standard net/http library except that it would
some extra lines of code.
I thought the extra lines of code was worth it
rather than having to maintain a single dependency.

shinenelson commented:
I don't know whether this repository is still in maintenance, but I had a use-case for this bot that made me come back to it. Upon navigating the repository is when I remembered that I hadn't worked on the changes that @jbernardo95 requested here. I've now made the changes that were requested.

Plus, I refactored away the `HTTParty` gem dependency that was kind of redundant because ruby already has the `net/http` standard library built-in. Besides, it was the only dependency and since there's only one outbound HTTP request, it wasn't too much of work to modify the request to work with the `net/http` standard library.
Of course, it's extra lines of code, but I think it's worth it so that we can remove one redundant dependency ( and the Gemfile altogether ).
